v2/internal/reconcilers/arm/arm_client.go

```diff
@@ -0,0 +1,166 @@
+	"github.com/Azure/azure-service-operator/v2/internal/config"
+	"github.com/Azure/azure-service-operator/v2/internal/genericarmclient"
+	"github.com/Azure/azure-service-operator/v2/internal/metrics"
+	"github.com/Azure/azure-service-operator/v2/internal/resolver"
+	"github.com/Azure/azure-service-operator/v2/internal/util/kubeclient"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
+	"github.com/pkg/errors"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	namespacedSecretName = "aso-credential"
+)
+
+type armClient struct {
+	genericClient  *genericarmclient.GenericClient
+	secretData     map[string][]byte
+	credentialFrom types.NamespacedName
+}
+
+func (c armClient) GenericClient() *genericarmclient.GenericClient {
+	return c.genericClient
+}
+
+func (c armClient) CredentialFrom() types.NamespacedName {
+	return c.credentialFrom
+}
+
+type armClients struct {
+	lock sync.Mutex
+	// clients map hold namespaced name for a secret to all the registered ARM clients
+	clients      map[string]*armClient
+	globalClient *armClient
+}
+
+func NewARMClients(client *genericarmclient.GenericClient, podNamespace string) *armClients {
+	globalClient := &armClient{
+		genericClient:  client,
+		credentialFrom: types.NamespacedName{Name: "aso-controller-settings", Namespace: podNamespace},
+	}
+
+	return &armClients{
+		lock:         sync.Mutex{},
+		clients:      make(map[string]*armClient),
+		globalClient: globalClient,
+	}
+}
+
+func newARMClient(client *genericarmclient.GenericClient, secretData map[string][]byte, credentialFrom types.NamespacedName) *armClient {
+	return &armClient{
+		genericClient:  client,
+		secretData:     secretData,
+		credentialFrom: credentialFrom,
+	}
+
+}
+
+func (c armClients) Register(client *armClient) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	c.clients[client.CredentialFrom().String()] = client
+}
+
+func (c armClients) Lookup(key string) (*armClient, bool) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	client, ok := c.clients[key]
+	return client, ok
+}
+
+func (c armClients) GetClient(ctx context.Context, obj genruntime.ARMMetaObject, kubeClient kubeclient.Client, cloud cloud.Configuration) (*armClient, error) {
+	// Namespaced secret
+	secret, err := getSecret(ctx, kubeClient, obj.GetNamespace(), namespacedSecretName)
+	if err != nil {
+		return nil, err
+	}
+
+	if secret != nil {
+		nsName := types.NamespacedName{Namespace: secret.Namespace, Name: secret.Name}
+		client, ok := c.Lookup(nsName.String())
+		// if client does not already exist, or secret data is changed, we need to get new token.
+		if !ok || !matchSecretData(client.secretData, secret.Data) {
+			credential, subscriptionID, err := c.newCredentialFromSecret(secret, nsName)
+			if err != nil {
+				return nil, err
+			}
+
+			newClient, err := genericarmclient.NewGenericClient(cloud, credential, subscriptionID, metrics.NewARMClientMetrics())
```

Comment:
Still need to pass `httpClient` here. (Resolving my other comment as it was out of date)

Reply:
done!